File: typescript/src/cli/json-keystore.ts
/**
 * Newline-delimited JSON file-backed {@link KeyStore} for the `licensing-keys` CLI.
 *
 * Why NDJSON and not a single JSON array? Atomic append of one line is
 * cheap with `fs.appendFileSync` + fsync; a whole-file rewrite loses work if
 * the process dies mid-write. On `put`/`update` we atomically rewrite via
 * write-to-tmp + rename, so a crash either preserves the old file or lands
 * the new one whole — never a half-written prefix.
 *
 * Schema: one LicenseKey JSON object per line, plus a sentinel first line
 * `{"__kind":"licensing-keys/v1"}` so we can evolve the format.
 */

import {
  closeSync,
  existsSync,
  fsyncSync,
  openSync,
  readFileSync,
  renameSync,
  writeFileSync,
} from 'node:fs';
import { dirname, resolve } from 'node:path';
import { errors } from '../errors.ts';
import type { KeyStore, KeyStoreFilter } from '../key-hierarchy.ts';
import type { LicenseKey, UUIDv7 } from '../types.ts';

const HEADER = '{"__kind":"licensing-keys/v1"}';

export class JsonFileKeyStore implements KeyStore {
  readonly #path: string;
  #byId = new Map<UUIDv7, LicenseKey>();
  #byKid = new Map<string, UUIDv7>();
  #loaded = false;

  constructor(path: string) {
    this.#path = resolve(path);
  }

  /** Force a read from disk. Called automatically by every mutating op. */
  #loadIfNeeded(): void {
    if (this.#loaded) return;
    if (!existsSync(this.#path)) {
      this.#loaded = true;
      return;
    }
    const text = readFileSync(this.#path, 'utf8');
    const lines = text.split('\n').filter((l) => l.length > 0);
    if (lines.length === 0) {
      this.#loaded = true;
      return;
    }
    const [first, ...rest] = lines;
    if (first !== HEADER) {
      throw errors.tokenMalformed(
        `keystore at ${this.#path}: missing or unknown header (expected ${HEADER})`,
      );
    }
    for (const line of rest) {
      const rec = JSON.parse(line) as LicenseKey;
      this.#byId.set(rec.id, rec);
      this.#byKid.set(rec.kid, rec.id);
    }
    this.#loaded = true;
  }

  /** Rewrite the file atomically: write-to-tmp then rename. */
  #flush(): void {
    const lines: string[] = [HEADER];
    for (const rec of this.#byId.values()) lines.push(JSON.stringify(rec));
    const tmp = `${this.#path}.tmp-${process.pid}-${Date.now()}`;
    writeFileSync(tmp, `${lines.join('\n')}\n`, { mode: 0o600 });
    // fsync the file before rename so the data is durable.
    const fd = openSync(tmp, 'r+');
    try {
      fsyncSync(fd);
    } finally {
      closeSync(fd);
    }
    renameSync(tmp, this.#path);
    // fsync the directory so the rename itself is durable on crash.
    try {
      const dir = openSync(dirname(this.#path), 'r');
      try {
        fsyncSync(dir);
      } finally {
        closeSync(dir);
      }
    } catch {
      // Directory fsync is best-effort; some filesystems don't support it.
    }
  }

  async put(record: LicenseKey): Promise<void> {
    this.#loadIfNeeded();
    const existingId = this.#byKid.get(record.kid);
    if (existingId !== undefined && existingId !== record.id) {
      throw errors.uniqueConstraintViolation('kid', record.kid);
    }
    this.#byId.set(record.id, record);
    this.#byKid.set(record.kid, record.id);
    this.#flush();
  }

  async get(id: UUIDv7): Promise<LicenseKey | null> {
    this.#loadIfNeeded();
    return this.#byId.get(id) ?? null;
  }

  async findByKid(kid: string): Promise<LicenseKey | null> {
    this.#loadIfNeeded();
    const id = this.#byKid.get(kid);
    return id ? (this.#byId.get(id) ?? null) : null;
  }

  async list(filter: KeyStoreFilter): Promise<readonly LicenseKey[]> {
    this.#loadIfNeeded();
    const out: LicenseKey[] = [];
    for (const rec of this.#byId.values()) {
      if (filter.scope_id !== undefined && rec.scope_id !== filter.scope_id) continue;
      if (filter.role !== undefined && rec.role !== filter.role) continue;
      if (filter.state !== undefined && rec.state !== filter.state) continue;
      if (filter.alg !== undefined && rec.alg !== filter.alg) continue;
      out.push(rec);
    }
    out.sort((a, b) => {
      if (a.created_at !== b.created_at) return a.created_at < b.created_at ? -1 : 1;
      return a.id < b.id ? -1 : a.id > b.id ? 1 : 0;
    });
    return out;
  }

  async update(id: UUIDv7, next: LicenseKey): Promise<void> {
    this.#loadIfNeeded();
    if (!this.#byId.has(id)) throw errors.tokenMalformed(`key not found: ${id}`);
    if (id !== next.id) throw errors.tokenMalformed('update cannot change id');
    this.#byId.set(id, next);
    this.#byKid.set(next.kid, id);
    this.#flush();
  }
}

File: cmd/licensing-keys/json_keystore.go
package main

// Newline-delimited JSON file-backed KeyStore for the licensing-keys CLI.
// Mirrors typescript/packages/core/src/cli/json-keystore.ts so a keystore
// file produced by one runtime can be consumed by the other — the wire
// format is `one LicenseKey JSON per line` with a leading sentinel:
//
//	{"__kind":"licensing-keys/v1"}
//	{"id":"...","kid":"...", ...}
//	{"id":"...","kid":"...", ...}
//	...
//
// Mutating ops rewrite atomically via write-to-tmp + rename so a crash
// either preserves the old file or lands the new one whole — never a
// half-written prefix. Directory fsync is best-effort (not all FSes
// support it).

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"sync"

	lic "github.com/AnoRebel/licensing/licensing"
)

const jsonStoreHeader = `{"__kind":"licensing-keys/v1"}`

// JSONFileKeyStore persists LicenseKey records to an NDJSON file.
// Concurrency: the struct is safe for concurrent use within a single
// process; multi-process contention is not protected (advisory file
// locking could be added if demand arises — the CLI is single-writer by
// design).
type JSONFileKeyStore struct {
	byID   map[string]lic.LicenseKey
	byKid  map[string]string
	path   string
	mu     sync.Mutex
	loaded bool
}

// NewJSONFileKeyStore constructs a JSONFileKeyStore bound to path. The
// file is created on first write.
func NewJSONFileKeyStore(path string) (*JSONFileKeyStore, error) {
	abs, err := filepath.Abs(path)
	if err != nil {
		return nil, err
	}
	return &JSONFileKeyStore{
		path:  abs,
		byID:  make(map[string]lic.LicenseKey),
		byKid: make(map[string]string),
	}, nil
}

func (s *JSONFileKeyStore) loadIfNeeded() error {
	if s.loaded {
		return nil
	}
	f, err := os.Open(s.path)
	if err != nil {
		if os.IsNotExist(err) {
			s.loaded = true
			return nil
		}
		return err
	}
	defer f.Close()

	scanner := bufio.NewScanner(f)
	// Allow long lines for RSA keys (encrypted PEM can push past 64 KB).
	scanner.Buffer(make([]byte, 0, 64*1024), 4*1024*1024)
	first := true
	for scanner.Scan() {
		line := strings.TrimRight(scanner.Text(), "\r")
		if line == "" {
			continue
		}
		if first {
			if line != jsonStoreHeader {
				return fmt.Errorf("keystore %s: missing or unknown header (expected %s)", s.path, jsonStoreHeader)
			}
			first = false
			continue
		}
		var rec lic.LicenseKey
		if err := json.Unmarshal([]byte(line), &rec); err != nil {
			return fmt.Errorf("keystore %s: malformed record: %w", s.path, err)
		}
		s.byID[rec.ID] = rec
		s.byKid[rec.Kid] = rec.ID
	}
	if err := scanner.Err(); err != nil {
		return err
	}
	s.loaded = true
	return nil
}

func (s *JSONFileKeyStore) flush() error {
	// Stable order: by CreatedAt then ID. Keeps diffs review-friendly.
	records := make([]lic.LicenseKey, 0, len(s.byID))
	for _, rec := range s.byID {
		records = append(records, rec)
	}
	sortRecords(records)

	lines := make([]string, 0, len(records)+1)
	lines = append(lines, jsonStoreHeader)
	for _, rec := range records {
		b, err := json.Marshal(rec)
		if err != nil {
			return err
		}
		lines = append(lines, string(b))
	}
	data := strings.Join(lines, "\n") + "\n"

	dir := filepath.Dir(s.path)
	if err := os.MkdirAll(dir, 0o700); err != nil {
		return err
	}
	tmp, err := os.CreateTemp(dir, filepath.Base(s.path)+".tmp-*")
	if err != nil {
		return err
	}
	tmpPath := tmp.Name()
	// On error, best-effort clean up.
	removeTmp := true
	defer func() {
		if removeTmp {
			_ = os.Remove(tmpPath)
		}
	}()
	if err := os.Chmod(tmpPath, 0o600); err != nil {
		tmp.Close()
		return err
	}
	if _, err := tmp.Write([]byte(data)); err != nil {
		tmp.Close()
		return err
	}
	// fsync file for durability.
	if err := tmp.Sync(); err != nil {
		tmp.Close()
		return err
	}
	if err := tmp.Close(); err != nil {
		return err
	}
	if err := os.Rename(tmpPath, s.path); err != nil {
		return err
	}
	removeTmp = false
	// Best-effort directory fsync (skipped silently if unsupported).
	if d, err := os.Open(dir); err == nil {
		_ = d.Sync()
		_ = d.Close()
	}
	return nil
}

// sortRecords orders by CreatedAt ascending, then ID ascending. Matches
// the TS emitter so diff-based review produces identical files across
// runtimes.
func sortRecords(in []lic.LicenseKey) {
	// Use a simple insertion sort tolerant of stable order; for n ≲ small
	// operational counts this is fine, and avoids importing sort.
	for i := 1; i < len(in); i++ {
		for j := i; j > 0 && lessRec(in[j], in[j-1]); j-- {
			in[j-1], in[j] = in[j], in[j-1]
		}
	}
}

func lessRec(a, b lic.LicenseKey) bool {
	if a.CreatedAt != b.CreatedAt {
		return a.CreatedAt < b.CreatedAt
	}
	return a.ID < b.ID
}

// KeyStore implementation ------------------------------------------------

func (s *JSONFileKeyStore) Put(rec lic.LicenseKey) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	if err := s.loadIfNeeded(); err != nil {
		return err
	}
	if existingID, ok := s.byKid[rec.Kid]; ok && existingID != rec.ID {
		return fmt.Errorf("UniqueConstraintViolation: kid %q already bound to id %s", rec.Kid, existingID)
	}
	s.byID[rec.ID] = rec
	s.byKid[rec.Kid] = rec.ID
	return s.flush()
}

func (s *JSONFileKeyStore) Get(id string) (*lic.LicenseKey, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if err := s.loadIfNeeded(); err != nil {
		return nil, err
	}
	rec, ok := s.byID[id]
	if !ok {
		return nil, nil
	}
	cp := rec
	return &cp, nil
}

func (s *JSONFileKeyStore) FindByKid(kid string) (*lic.LicenseKey, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if err := s.loadIfNeeded(); err != nil {
		return nil, err
	}
	id, ok := s.byKid[kid]
	if !ok {
		return nil, nil
	}
	rec := s.byID[id]
	return &rec, nil
}

func (s *JSONFileKeyStore) List(filter lic.KeyStoreFilter) ([]lic.LicenseKey, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if err := s.loadIfNeeded(); err != nil {
		return nil, err
	}
	out := make([]lic.LicenseKey, 0, len(s.byID))
	for _, rec := range s.byID {
		if filter.ScopeIDSet {
			if !scopeEq(rec.ScopeID, filter.ScopeID) {
				continue
			}
		}
		if filter.Role != nil && rec.Role != *filter.Role {
			continue
		}
		if filter.State != nil && rec.State != *filter.State {
			continue
		}
		if filter.Alg != nil && rec.Alg != *filter.Alg {
			continue
		}
		out = append(out, rec)
	}
	sortRecords(out)
	return out, nil
}

func (s *JSONFileKeyStore) Update(id string, next lic.LicenseKey) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	if err := s.loadIfNeeded(); err != nil {
		return err
	}
	if _, ok := s.byID[id]; !ok {
		return fmt.Errorf("TokenMalformed: key not found: %s", id)
	}
	if id != next.ID {
		return fmt.Errorf("TokenMalformed: update cannot change id")
	}
	s.byID[id] = next
	s.byKid[next.Kid] = id
	return s.flush()
}

func scopeEq(a, b *string) bool {
	if a == nil && b == nil {
		return true
	}
	if a == nil || b == nil {
		return false
	}
	return *a == *b
}
